File: sdks/java/src/test/java/dev/atap/sdk/api/CredentialApiTest.java
package dev.atap.sdk.api;

import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import dev.atap.sdk.ATAPClient;
import dev.atap.sdk.TestFixtures;
import dev.atap.sdk.model.Credential;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static org.junit.jupiter.api.Assertions.*;

class CredentialApiTest {

    private WireMockServer wireMock;
    private ATAPClient client;

    @BeforeEach
    void setUp() {
        wireMock = new WireMockServer(WireMockConfiguration.wireMockConfig().dynamicPort());
        wireMock.start();

        wireMock.stubFor(post(urlEqualTo("/v1/oauth/token"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.tokenResponseJson())));

        client = ATAPClient.builder()
                .baseUrl("http://localhost:" + wireMock.port())
                .did(TestFixtures.TEST_DID)
                .keyPair(TestFixtures.getTestKeyPair())
                .clientSecret(TestFixtures.TEST_CLIENT_SECRET)
                .platformDomain(TestFixtures.TEST_PLATFORM_DOMAIN)
                .build();
    }

    @AfterEach
    void tearDown() {
        wireMock.stop();
        client.close();
    }

    @Test
    void testStartEmail() {
        wireMock.stubFor(post(urlEqualTo("/v1/credentials/email/start"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody("{\"message\":\"OTP sent to test@example.com\"}")));

        String result = client.credentials.startEmail("test@example.com");
        assertEquals("OTP sent to test@example.com", result);
    }

    @Test
    void testVerifyEmail() {
        wireMock.stubFor(post(urlEqualTo("/v1/credentials/email/verify"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.credentialResponseJson())));

        Credential cred = client.credentials.verifyEmail("test@example.com", "123456");
        assertNotNull(cred);
        assertEquals("cred_abc123", cred.getId());
        assertEquals("ATAPEmailVerification", cred.getType());
    }

    @Test
    void testStartPhone() {
        wireMock.stubFor(post(urlEqualTo("/v1/credentials/phone/start"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody("{\"message\":\"OTP sent\"}")));

        String result = client.credentials.startPhone("+15551234567");
        assertEquals("OTP sent", result);
    }

    @Test
    void testVerifyPhone() {
        wireMock.stubFor(post(urlEqualTo("/v1/credentials/phone/verify"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.credentialResponseJson())));

        Credential cred = client.credentials.verifyPhone("+15551234567", "654321");
        assertNotNull(cred);
    }

    @Test
    void testSubmitPersonhood() {
        wireMock.stubFor(post(urlEqualTo("/v1/credentials/personhood"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.credentialResponseJson())));

        Credential cred = client.credentials.submitPersonhood("provider_token_123");
        assertNotNull(cred);
        assertEquals("cred_abc123", cred.getId());
    }

    @Test
    void testSubmitPersonhoodWithoutToken() {
        wireMock.stubFor(post(urlEqualTo("/v1/credentials/personhood"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(TestFixtures.credentialResponseJson())));

        Credential cred = client.credentials.submitPersonhood();
        assertNotNull(cred);
    }

    @Test
    void testList() {
        String listJson = "{\"credentials\":[" + TestFixtures.credentialResponseJson() + "]}";

        wireMock.stubFor(get(urlEqualTo("/v1/credentials"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody(listJson)));

        List<Credential> creds = client.credentials.list();
        assertNotNull(creds);
        assertEquals(1, creds.size());
        assertEquals("cred_abc123", creds.get(0).getId());
    }

    @Test
    void testStatusList() {
        wireMock.stubFor(get(urlEqualTo("/v1/credentials/status/1"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody("{\"type\":\"BitstringStatusListCredential\"}")));

        Map<String, Object> result = client.credentials.statusList();
        assertEquals("BitstringStatusListCredential", result.get("type"));
    }

    @Test
    void testStatusListCustomId() {
        wireMock.stubFor(get(urlEqualTo("/v1/credentials/status/42"))
                .willReturn(aResponse()
                        .withStatus(200)
                        .withHeader("Content-Type", "application/json")
                        .withBody("{\"type\":\"BitstringStatusListCredential\"}")));

        Map<String, Object> result = client.credentials.statusList("42");
        assertNotNull(result);
    }
}

File: sdks/php/tests/Api/CredentialApiTest.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Tests\Api;

use Atap\Sdk\Model\Credential;
use Atap\Sdk\Tests\TestCase;

class CredentialApiTest extends TestCase
{
    public function testStartEmail(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, ['message' => 'OTP sent to user@example.com']),
        ]);

        $message = $client->credentials()->startEmail('user@example.com');
        $this->assertSame('OTP sent to user@example.com', $message);
    }

    public function testVerifyEmail(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'id' => 'cred_123',
                'type' => 'ATAPEmailVerification',
                'credential' => 'eyJhbGciOi...',
                'issued_at' => '2024-01-01T00:00:00Z',
            ]),
        ]);

        $cred = $client->credentials()->verifyEmail('user@example.com', '123456');
        $this->assertInstanceOf(Credential::class, $cred);
        $this->assertSame('cred_123', $cred->id);
        $this->assertSame('ATAPEmailVerification', $cred->type);
    }

    public function testStartPhone(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, ['message' => 'OTP sent']),
        ]);

        $message = $client->credentials()->startPhone('+1234567890');
        $this->assertSame('OTP sent', $message);
    }

    public function testVerifyPhone(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'id' => 'cred_456',
                'type' => 'ATAPPhoneVerification',
                'credential' => 'eyJhbGciOi...',
                'issued_at' => '2024-01-01T00:00:00Z',
            ]),
        ]);

        $cred = $client->credentials()->verifyPhone('+1234567890', '654321');
        $this->assertInstanceOf(Credential::class, $cred);
        $this->assertSame('ATAPPhoneVerification', $cred->type);
    }

    public function testSubmitPersonhood(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'id' => 'cred_789',
                'type' => 'ATAPPersonhood',
                'credential' => 'eyJhbGciOi...',
                'issued_at' => '2024-01-01T00:00:00Z',
            ]),
        ]);

        $cred = $client->credentials()->submitPersonhood('provider_token_abc');
        $this->assertSame('ATAPPersonhood', $cred->type);
    }

    public function testSubmitPersonhoodNoToken(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'id' => 'cred_no_token',
                'type' => 'ATAPPersonhood',
                'credential' => 'eyJhbGciOi...',
                'issued_at' => '2024-01-01T00:00:00Z',
            ]),
        ]);

        $cred = $client->credentials()->submitPersonhood();
        $this->assertSame('cred_no_token', $cred->id);
    }

    public function testListCredentials(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, [
                'credentials' => [
                    ['id' => 'cred_1', 'type' => 'ATAPEmailVerification', 'credential' => 'jwt1', 'issued_at' => '2024-01-01T00:00:00Z'],
                    ['id' => 'cred_2', 'type' => 'ATAPPhoneVerification', 'credential' => 'jwt2', 'issued_at' => '2024-01-02T00:00:00Z'],
                ],
            ]),
        ]);

        $creds = $client->credentials()->list();
        $this->assertCount(2, $creds);
        $this->assertSame('cred_1', $creds[0]->id);
    }

    public function testStatusList(): void
    {
        $client = $this->createUnauthClient([
            $this->jsonResponse(200, [
                'id' => 'https://example.com/credentials/status/1',
                'type' => ['VerifiableCredential', 'BitstringStatusListCredential'],
            ]),
        ]);

        $data = $client->credentials()->statusList('1');
        $this->assertArrayHasKey('id', $data);
    }

    public function testStartEmailDefaultMessage(): void
    {
        $client = $this->createAuthedClient([
            $this->jsonResponse(200, []),
        ]);

        $message = $client->credentials()->startEmail('test@test.com');
        $this->assertSame('OTP sent', $message);
    }
}
